c2/cuda/groth16_cuda.cu

```diff
@@ -280,14 +292,36 @@ RustError::by_value generate_groth16_proofs_c(const Assignment<fr_t> provers[],
         if (caught_exception)
             return;
 
-        split_vectors_l.tail_msms_resize(l_popcount);
-        split_vectors_a.tail_msms_resize(a_popcount);
-        split_vectors_b.tail_msms_resize(b_popcount);
-
-        tail_msm_l_bases.resize(l_popcount);
-        tail_msm_a_bases.resize(a_popcount);
-        tail_msm_b_g1_bases.resize(b_popcount);
-        tail_msm_b_g2_bases.resize(b_popcount);
+        if (l_popcount >= (float)0.5 * points_l.size())
+            l_skip_batch_add = true;
+        if (a_popcount >= (float)0.5 * points_a.size())
+            a_skip_batch_add = true;
+        if (b_popcount >= (float)0.5 * points_b_g1.size())
+            b_skip_batch_add = true;
+
+        if (l_skip_batch_add)
+            split_vectors_l.tail_msms_resize(points_l.size());
+        else
+            split_vectors_l.tail_msms_resize(l_popcount);
+
+        if (a_skip_batch_add)
+            split_vectors_a.tail_msms_resize(points_a.size());
+        else
+            split_vectors_a.tail_msms_resize(a_popcount);
+
+        if (b_skip_batch_add)
+            split_vectors_b.tail_msms_resize(points_b_g1.size());
+        else
+            split_vectors_b.tail_msms_resize(b_popcount);
+
+        if (!l_skip_batch_add)
+            tail_msm_l_bases.resize(l_popcount);
```

Comment:
Just in case for reference, for example as
```
        if (l_skip_batch_add = (l_popcount >= points_l.size()/2)) {
            split_vectors_l.tail_msms_resize(points_l.size());
        } else {
            split_vectors_l.tail_msms_resize(l_popcount);
            tail_msm_l_bases.resize(l_popcount);
        }
```

Reply:
This naturally applies to `a` and `b`. While specifically about `l`. Does one actually have to copy the scalars? I mean can't one simply pass the aux_assignment_data in the l_skip_batch_add case?

Reply:
> I mean can't one simply pass the aux_assignment_data in the l_skip_batch_add case?

One can indeed do this!